Use raft state in placement

youngbupark opened the pull request:
# Description

This PR is to use the state (dapr runtime host list and consistent hashing table) maintained by Raft consensus algorithm and include following improvements:

1. Batched table dissemination instead of disseminate hashing tables to runtimes immediately.
   - improve runtime and placement service stability.
2. Split runtime connection from runtime membership.
   - When there is intermittent failures between placement and Dapr runtime, placement will not delete host from the membership and update consistent hashing table immediately.
3. Whenever runtime sends heartbeat to placement, placement records the latest heartbeat timestamp. 
   - Placement will remove the member from membership and update tables only when the last heartbeat timestamp is outdated. But if runtime disconnects the connection from placement gracefully, then it will remove the member from membership in the next flush tick. 

Next work:
- [ ] dapr runtime will disconnect the connection from placement gracefully when it gets exit signal.
- [ ] dapr runtime can connect to leader of multiple placement instance 
- [ ] helm chart update.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #663 #2089

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

youngbupark commented:
/ok-to-test

youngbupark commented on pkg/placement/placement.go:
```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-}
-
-// ProcessRemovedHost removes a host from the hash table
-func (p *Service) ProcessRemovedHost(id string) {
```
This is no longer required because raft FSM logic handles the hashing table updates.

youngbupark commented on pkg/placement/placement.go:
```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-
-// PerformTablesUpdate updates the connected dapr runtimes using a 3 stage commit. first it locks so no further dapr can be taken
-// it then proceeds to update and then unlock once all runtimes have been updated
-func (p *Service) PerformTablesUpdate(hosts []placementv1pb.Placement_ReportDaprStatusServer,
```
this function is refactored to remove all duplicated codes in membership.go. I added unit-tests to validate lock/update/unlock stages.

youngbupark commented on pkg/placement/placement.go:
```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
 }
 
-// ProcessHost updates the distributed has list based on a new host and its entities
-func (p *Service) ProcessHost(host *placementv1pb.Host) {
```
This logic is moved to FSM state logic, similar to ProcessRemovedHost.

youngbupark commented:
@yaron2 @artursouza all missing test for the existing code, such as gRPC stream connection/disconnection and three phase updates, are added.

artursouza commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+			case raft.MemberRemove:
+				updated, raftErr := p.raftNode.ApplyCommand(op.cmdType, op.host)
+				if raftErr != nil {
+					log.Debugf("fail to apply command: %v", raftErr)
```
Shouldn't the worker restart with a clean state in case of error?

youngbupark replied:
+1 worker should not restart. instead, we may need to retry apply command.

youngbupark replied:
I missed why I didn't add retry here :) each runtime will send the heartbeat every 1 second. For MemberUpsert, as long as runtime sends heartbeat, it will try to add or update updateAt timestamp in Raft state every 1 second. 

For MemberRemove, the faulty host detect timer at L65 will try to remove node if MemberRemove was not successful.

so the state will be eventually consistent. but I agree to add simple linear retry here.

youngbupark replied:
added it.

artursouza commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+	p.disseminateLock.Lock()
+	defer p.disseminateLock.Unlock()
+
+	p.disseminateOperation(hosts, "lock", nil)
```
Return in case of error.

youngbupark replied:
Same reason above. I would like to use the existing behavior. instead, I would like to address it in the separate issue.

artursouza commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+				}
+
+				log.Debugf("desseminate tables to runtimes. request count: %d", hostUpdateCount)
+				p.performTablesUpdate(p.streamConns, p.raftNode.FSM().PlacementState())
```
handle error here too.

youngbupark replied:
The existing table update doesn't care the error. I would not change the existing behavior in this PR. Instead, I will address it by the separated issue.

youngbupark replied:
I will add TODO comment here.

youngbupark replied:
added comment

youngbupark commented:
/ok-to-test

youngbupark commented on pkg/placement/placement.go:
```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-	v := fmt.Sprintf("%v", p.generation)
-
-	o.Operation = "update"
-	o.Tables = &placementv1pb.PlacementTables{
```
this part is moved to `PlacementState()` in fsm.go.

mukundansundar commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// performTablesUpdate updates the connected dapr runtimes using a 3 stage commit.
+// first it locks so no further dapr can be taken it then proceeds to update and
```
nitpick phrasing ...

youngbupark replied:
I copied from the exsiting code :) I rephrased it as possible as I can.

mukundansundar commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+// MembershipChangeWorker is the worker to change the state of membership
+// and update the consistent hashing tables for actors.
+func (p *Service) MembershipChangeWorker() {
+	faultHostDetectTimer := time.NewTicker(faultyHostDetectInterval)
```
nitpick: `faultyHostDetectTimer`

youngbupark replied:
fixed.

mukundansundar commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,118 @@
+		err = s.Send(o)
+		if err != nil {
+			log.Errorf("error updating host on unlock operation: %s", err)
+			// TODO: the error should not be ignored. By handing error or retrying dissemination,
```
is this TODO still valid and will it be fixed as part of this PR?

youngbupark replied:
nope. I will keep using the existing behavior and address this by the separate issue. I want to address the existing logic's problem first and note that we need to improve it later.

mukundansundar commented on pkg/placement/membership_test.go:
```diff
@@ -0,0 +1,150 @@
+	// act
+	for i := 0; i < testClients; i++ {
+		host := &v1pb.Host{
+			Name:     fmt.Sprintf("127.0.0.1:5010%d", i),
```
would it be better if freeport is used instead of deterministically setting a port?

youngbupark replied:
this is the fake address :) it will not occupy any port.

mukundansundar commented on pkg/placement/membership_test.go:
```diff
@@ -0,0 +1,150 @@
+		assert.NoError(t, clientStreams[i].Send(host))
+	}
+
+	time.Sleep(10 * time.Millisecond)
```
comment on why this is needed.

youngbupark replied:
added

mukundansundar commented on pkg/placement/membership_test.go:
```diff
@@ -0,0 +1,150 @@
+
+	time.Sleep(10 * time.Millisecond)
+
+	testServer.performTablesUpdate(testServer.streamConns, nil)
```
can we just wait for flushTimerInterval like the previous test? Why is this called directly? And why is the tables argument nil?

youngbupark replied:
this test is unit-test to validate the behavior of only performTableUpdate. so I directly call performTableUpdate without dependency of worker loop.

mukundansundar commented on pkg/placement/placement.go:
```diff
@@ -25,42 +24,99 @@ import (
+	// it updates the last heartbeat time in UpdateAt of the FSM state. If Now - UpdatedAt exceeds
+	// faultyHostDetectMaxDuration, MembershipChangeWorker tries to remove faulty dapr runtime from
+	// membership.
+	faultyHostDetectMaxDuration = 3 * time.Second
```
not needed for this PR possibly, but can all the times be made as configurable parameters?

youngbupark replied:
may be later.

mukundansundar commented on pkg/placement/placement.go:
```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-			monitoring.RecordPerActorTypeReplicasCount(e, host.Name)
+func (p *Service) deleteRuntimeConnection(conn placementGRPCStream) {
+	p.streamConnsLock.Lock()
+	for i, c := range p.streamConns {
```
should streamConns be a map for faster lookup? Since there might be the scenario that there are multiple connections being added and removed for actors frequently.

mukundansundar commented on pkg/placement/placement_test.go:
```diff
@@ -0,0 +1,178 @@
+	testRaftServer = raft.New("testnode", true, true, []raft.PeerInfo{
+		{
+			ID:      "testnode",
+			Address: "127.0.0.1:6060",
```
do we know that 6060 port is always free?

youngbupark replied:
this is fake port, this code doesn't occupy any port.

mukundansundar commented on pkg/placement/placement_test.go:
```diff
@@ -0,0 +1,178 @@
+	testRaftServer.StartRaft(nil)
+
+	for {
+		if testRaftServer.IsLeader() {
```
is this added to make sure that the server starts up correctly? Please add comment if so.

youngbupark replied:
yes this is to wait until raft framework selects the current test raft server as leader.

youngbupark replied:
added comment

mukundansundar commented on pkg/placement/placement_test.go:
```diff
@@ -0,0 +1,178 @@
+			assert.EqualValues(t, host.Entities, memberChange.host.Entities)
+			assert.Equal(t, 1, len(testServer.streamConns))
+
+		case <-time.After(100 * time.Millisecond):
```
should this be after flushTimer?

youngbupark replied:
no this is not related to flushTimer. This test is testing for ReportDaprStatus method. we can put any arbitrary number but I choose 100 millisecond to wait until ReportDaprStatus api queued member change to membershipCh channel.

youngbupark replied:
I defined constant to clarify the timeout usage.

mukundansundar commented on pkg/placement/placement_test.go:
```diff
@@ -0,0 +1,178 @@
+			require.True(t, false, "should not have any member change message because faulty host detector time will clean up")
+
+		case <-time.After(100 * time.Millisecond):
+			assert.Equal(t, 0, len(testServer.streamConns))
```
should this be after faultyHostTimer?

youngbupark replied:
same above.

mukundansundar commented on pkg/placement/raft/fsm_test.go:
```diff
@@ -34,10 +34,11 @@ func TestFSMApply(t *testing.T) {
 		}
 
 		resp := fsm.Apply(raftLog)
```
should Apply return bool, error instead of interface{}?

youngbupark replied:
this is the by-design of hashicorp raft implementation. resp can be error or actual value. so it always needs to check if resp can be type conversion or not. please look at the other code to see how it handles error or bool.

youngbupark commented:
@artursouza @mukundansundar I addressed all your comments. please review it again. the PR aims at using raft state with smaller change in the existing placement behavior, such as `performTableUpdate`. But I tried to refactor the existing code to make it testable and added comments and unit-tests. So, `TODO` comment will be addressed by the separate issues and PRs.

youngbupark commented:
/ok-to-test

mukundansundar commented:
/ok-to-test

youngbupark commented on pkg/placement/placement.go:
```diff
@@ -25,42 +24,102 @@ import (
-// ReportDaprStatus gets a heartbeat report from different Dapr hosts
+// Shutdown close all server connections.
+func (p *Service) Shutdown() {
+	if p.grpcServer != nil {
```
no need to close channel explcitly. it will be GCed.

youngbupark commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,146 @@
+
+	for {
+		if !p.raftNode.IsLeader() {
+			// TODO: use leader election channel instead of sleep
```
this will be done by https://github.com/dapr/dapr/issues/2375

youngbupark commented:
/ok-to-test

yaron2 commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,146 @@
+
+	for {
+		if !p.raftNode.IsLeader() {
+			// TODO: use leader election channel instead of sleep
```
What stops us from using the election channel for this PR?

youngbupark replied:
As I explained in @mukundansundar 's question, this will be done by next PR. This code will be ignored for the current single instance scenario.

yaron2 commented on pkg/placement/membership.go:
```diff
@@ -0,0 +1,146 @@
+		if !p.raftNode.IsLeader() {
+			// TODO: use leader election channel instead of sleep
+			// <- p.raftNode.leaderCh
+			time.Sleep(100 * time.Millisecond)
```
Can this be a const?

youngbupark replied:
I will get rid of this code in the next PR.

yaron2 reviewed (approved):
LGTM
